src/types/rule.cpp

```diff
@@ -332,18 +386,40 @@ void Rule::removeString(const std::string& id)
  */
 void Rule::addTag(const std::string& tag)
 {
-	_tags.push_back(tag);
+	//find iterator behind tags in TokenStream
+	TokenIt end = ++_tags.back();
+	TokenIt newTagIt = _tokenStream->insert(end, TokenType::TAG, Literal(tag));
+	_tags.push_back(newTagIt);
 }
 
 /**
  * Removes tag from the rule.
  *
  * @param tag Tag to remove.
+ * @return true iff there was corresponding tag and it was removed
  */
 void Rule::removeTags(const std::string& tag)
 {
-	auto new_end = std::remove(_tags.begin(), _tags.end(), tag);
-	_tags.erase(new_end, _tags.end());
+	for (auto it = _tags.begin(); it != _tags.end(); ++it)
+		if ((*it)->getText() == tag)
+		{
+			_tokenStream->erase(*it);
+			_tags.erase(it);
```

Comment:
It is not safe to iterate over vector while erasing from it. Read the description of `erase()` method on cppreference.

> Invalidates iterators and references at or after the point of the erase, including the end() iterator. 

https://en.cppreference.com/w/cpp/container/vector/erase

Reply:
Ok, I used `std::find_if` to find the iterator and then use it to remove the tag.